<a href="https://colab.research.google.com/github/harrshat/MachineLearningReferenceDocs/blob/main/Used%20Car%20Price%20Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Used Car Price Prediction

- This dataset contains used car prices sold on cardekho.com in India as well as more information about the cars.
- User - predicts price of the car based on what we know about the car specs
- Prediction results can be used by the new seller to get an idea of the price of the car based on the market.

- AIM: Predict the price of the car - Continuous - Regression

In [73]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [74]:
df=pd.read_csv("/content/cardekho_imputated.csv")

In [75]:
df.head(5)

,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000


In [76]:
df.shape

(15411, 14)

In [77]:
df.isnull().sum()
# Inference - No nulls

,0
Unnamed: 0,0
car_name,0
brand,0
model,0
vehicle_age,0
km_driven,0
seller_type,0
fuel_type,0
transmission_type,0
mileage,0


In [78]:
# Drop the first column
df.drop('Unnamed: 0',axis=1,inplace=True)

In [79]:
df.shape

(15411, 13)

In [80]:
df.head(3)

,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.7,796,46.3,5,120000
1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.9,1197,82.0,5,550000
2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.0,1197,80.0,5,215000


In [81]:
# Drop unnecessary columns. Here, car_name = brand + model. One column (model) is sufficient.
df.drop('car_name',axis=1,inplace=True)
df.drop('brand',axis=1,inplace=True)

In [82]:
df.head(3)

,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,Alto,9,120000,Individual,Petrol,Manual,19.7,796,46.3,5,120000
1,Grand,5,20000,Individual,Petrol,Manual,18.9,1197,82.0,5,550000
2,i20,11,60000,Individual,Petrol,Manual,17.0,1197,80.0,5,215000


In [83]:
df.shape

(15411, 11)

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15411 entries, 0 to 15410
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   model              15411 non-null  object 
 1   vehicle_age        15411 non-null  int64  
 2   km_driven          15411 non-null  int64  
 3   seller_type        15411 non-null  object 
 4   fuel_type          15411 non-null  object 
 5   transmission_type  15411 non-null  object 
 6   mileage            15411 non-null  float64
 7   engine             15411 non-null  int64  
 8   max_power          15411 non-null  float64
 9   seats              15411 non-null  int64  
 10  selling_price      15411 non-null  int64  
dtypes: float64(2), int64(5), object(4)
memory usage: 1.3+ MB


In [85]:
#df.shape--(15411, 11)-- before dropping duplicates
df=df[~df.duplicated()]
## Inference- 167 Duplicate rows dropped
df.shape #(15244, 11)

(15244, 11)

In [86]:
cat_features=[col for col in df.columns if df[col].dtype=='object']
num_features=[col for col in df.columns if df[col].dtype!='object']
print("Categorical Features:",cat_features)
print("Numerical Features:",num_features)

Categorical Features: ['model', 'seller_type', 'fuel_type', 'transmission_type']
Numerical Features: ['vehicle_age', 'km_driven', 'mileage', 'engine', 'max_power', 'seats', 'selling_price']


In [87]:
df.describe()

,vehicle_age,km_driven,mileage,engine,max_power,seats,selling_price
count,15244.000000,1.524400e+04,15244.000000,15244.000000,15244.000000,15244.000000,1.524400e+04
mean,6.041131,5.563958e+04,19.697333,1486.171543,100.607652,5.326161,7.747014e+05
std,3.016228,5.176630e+04,4.169307,520.419390,42.915687,0.808760,8.946761e+05
min,0.000000,1.000000e+02,4.000000,793.000000,38.400000,0.000000,4.000000e+04
25%,4.000000,3.000000e+04,17.000000,1197.000000,74.000000,5.000000,3.850000e+05
50%,6.000000,5.000000e+04,19.670000,1248.000000,88.500000,5.000000,5.590000e+05
75%,8.000000,7.000000e+04,22.700000,1582.000000,117.300000,5.000000,8.250000e+05
max,29.000000,3.800000e+06,33.540000,6592.000000,626.000000,9.000000,3.950000e+07


In [88]:
for feature in cat_features:
  print(feature)
  print(df[feature].value_counts())
  print("========================================")

model
model
i20             898
Swift Dzire     875
Swift           774
Alto            768
City            750
               ... 
Altroz            1
C                 1
Ghost             1
Quattroporte      1
Gurkha            1
Name: count, Length: 120, dtype: int64
seller_type
seller_type
Dealer              9459
Individual          5612
Trustmark Dealer     173
Name: count, dtype: int64
fuel_type
fuel_type
Petrol      7555
Diesel      7342
CNG          299
LPG           44
Electric       4
Name: count, dtype: int64
transmission_type
transmission_type
Manual       12094
Automatic     3150
Name: count, dtype: int64


In [89]:
X=df.drop('selling_price',axis=1)
y=df['selling_price']

In [90]:
from sklearn.model_selection import train_test_split

In [91]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [92]:
X_train.shape,y_train.shape

((12195, 10), (12195,))

In [93]:
X_test.shape,y_test.shape

((3049, 10), (3049,))

In [94]:
X.head(3)

,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats
0,Alto,9,120000,Individual,Petrol,Manual,19.7,796,46.3,5
1,Grand,5,20000,Individual,Petrol,Manual,18.9,1197,82.0,5
2,i20,11,60000,Individual,Petrol,Manual,17.0,1197,80.0,5


# Feature Scaling and Encoding
- Label Encoding for 'model' column and not OneHotEncoding since there are 120 unique values. OneHotEncoding for columns which have less unique values and not ordinal

In [95]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
X['model']=le.fit_transform(X['model'])

In [96]:
X.head()

,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats
0,7,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5
1,54,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5
2,118,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5
3,7,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5
4,38,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5


In [98]:
df['seller_type'].unique()

array(['Individual', 'Dealer', 'Trustmark Dealer'], dtype=object)

In [99]:
df['fuel_type'].unique()

array(['Petrol', 'Diesel', 'CNG', 'LPG', 'Electric'], dtype=object)

In [100]:
df['transmission_type'].unique()

array(['Manual', 'Automatic'], dtype=object)

In [101]:
df.columns

Index(['model', 'vehicle_age', 'km_driven', 'seller_type', 'fuel_type',
       'transmission_type', 'mileage', 'engine', 'max_power', 'seats',
       'selling_price'],
      dtype='object')

In [104]:
# Create Column transformer using 3 different transformers
num_features=X.select_dtypes(exclude='object').columns
print(num_features)
onehot_columns=['seller_type','fuel_type','transmission_type']
from sklearn.preprocessing import OneHotEncoder,StandardScaler
from sklearn.compose import ColumnTransformer

Index(['model', 'vehicle_age', 'km_driven', 'mileage', 'engine', 'max_power',
       'seats'],
      dtype='object')


In [109]:
numeric_transformer=StandardScaler()
cat_transformer=OneHotEncoder(drop='first')

preprocessor = ColumnTransformer(
    [
        ("OneHotEncoder",cat_transformer,onehot_columns),
        ("StandardScaler",numeric_transformer,num_features)
    ],remainder='passthrough' #other than these columns, the rest of the columns will stay
)

In [110]:
print(preprocessor)

ColumnTransformer(remainder='passthrough',
                  transformers=[('OneHotEncoder', OneHotEncoder(drop='first'),
                                 ['seller_type', 'fuel_type',
                                  'transmission_type']),
                                ('StandardScaler', StandardScaler(),
                                 Index(['model', 'vehicle_age', 'km_driven', 'mileage', 'engine', 'max_power',
       'seats'],
      dtype='object'))])


In [113]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [115]:
X_train.head()

,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats
3809,42,5,41974,Dealer,Diesel,Manual,24.52,1248,88.50,7
4848,100,16,90000,Individual,Petrol,Manual,14.00,1061,64.00,5
1525,95,10,80000,Individual,Diesel,Manual,20.54,1598,103.60,5
13708,87,8,55572,Dealer,Petrol,Automatic,13.70,1798,157.75,5
4557,64,4,110000,Individual,Petrol,Manual,23.01,999,67.00,5


In [116]:
X_train=preprocessor.fit_transform(X_train)
X_test=preprocessor.transform(X_test)

In [121]:
pd.DataFrame(X_train).head()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13
0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,-0.551812,-0.347650,-0.313297,1.166383,-0.450611,-0.278238,2.093957
1,1.0,0.0,0.0,0.0,0.0,1.0,1.0,1.044719,3.285533,0.823151,-1.368036,-0.809634,-0.844091,-0.401480
2,1.0,0.0,1.0,0.0,0.0,0.0,1.0,0.907087,1.303797,0.586519,0.207544,0.221358,0.070513,-0.401480
3,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.686876,0.643218,0.008475,-1.440310,0.605341,1.321164,-0.401480
4,1.0,0.0,0.0,0.0,0.0,1.0,1.0,0.053768,-0.677940,1.296414,0.802603,-0.928669,-0.774803,-0.401480


# Model Training and Model Selection

In [123]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error

In [124]:
# Create a function for Model Selection and Training
def evaluate_model(true,predicted):
  mae=mean_absolute_error(true,predicted)
  mse=mean_squared_error(true,predicted)
  rmse=np.sqrt(mean_squared_error(true,predicted))
  r2score=r2_score(true,predicted)
  return mae,rmse,r2score


In [126]:
# Beginning Model Training
models={
    "RandomForest":RandomForestRegressor(),
    "LinearRegression":LinearRegression(),
    "KNN":KNeighborsRegressor(),
    "DecisionTree":DecisionTreeRegressor(),
    "Ridge":Ridge(),
    "Lasso":Lasso()
}

In [127]:
print(models)

{'RandomForest': RandomForestRegressor(), 'LinearRegression': LinearRegression(), 'KNN': KNeighborsRegressor(), 'DecisionTree': DecisionTreeRegressor(), 'Ridge': Ridge(), 'Lasso': Lasso()}


In [140]:
for i in range(len(list(models))):
  model=list(models.values())[i]
  model.fit(X_train,y_train)

  # For training data
  y_train_pred=model.predict(X_train)
  y_test_pred=model.predict(X_test)

  print(model)
  X_train_mae,X_train_rmse,X_train_r2score=evaluate_model(y_train,y_train_pred)
  X_test_mae,X_test_rmse,X_test_r2score=evaluate_model(y_test,y_test_pred)

  print(list(models.keys())[i])
  print("---------------Train---------------")
  print("Mean Absolute Error:")
  print(X_train_mae)
  print("Root Mean Square Error:")
  print(X_train_rmse)
  print("R2 Score:")
  print(X_train_r2score)
  print("---------------Test---------------")
  print("Mean Absolute Error:")
  print(X_test_mae)
  print("Root Mean Square Error:")
  print(X_test_rmse)
  print("R2 Score:")
  print(X_test_r2score)

  print("===========================================================================")


RandomForestRegressor()
RandomForest
---------------Train---------------
Mean Absolute Error:
40087.92996141938
Root Mean Square Error:
125938.63294754918
R2 Score:
0.9813073225759107
---------------Test---------------
Mean Absolute Error:
102700.06813330797
Root Mean Square Error:
299036.11854549847
R2 Score:
0.8529283920020249
LinearRegression()
LinearRegression
---------------Train---------------
Mean Absolute Error:
275284.3899581493
Root Mean Square Error:
566013.4712838263
R2 Score:
0.6224216387180522
---------------Test---------------
Mean Absolute Error:
262717.18861080665
Root Mean Square Error:
446765.18534247123
R2 Score:
0.6717231990253707
KNeighborsRegressor()
KNN
---------------Train---------------
Mean Absolute Error:
91454.63509635096
Root Mean Square Error:
301039.5757603898
R2 Score:
0.8931927462499339
---------------Test---------------
Mean Absolute Error:
112497.85995408331
Root Mean Square Error:
321025.9116381995
R2 Score:
0.8305031144071374
DecisionTreeRegressor(

## Conclusion - Random Forest Regressor Model and KNN gave the best results
- Further, hyper parameter tuning can be done by changing different parameters for each model separately to improve model performance.